File: typescript/src/terminals.ts
import { Clause } from './clause.js';
import { Match, mismatch, type MatchResult } from './matchResult.js';
import type { Parser } from './parser.js';
import { escapeString } from './utils.js';

// -----------------------------------------------------------------------------------------------------------------

/**
 * Abstract base class for terminal clauses.
 */
export abstract class Terminal extends Clause {
  /** The AST/CST node label for terminals. */
  static readonly nodeLabel = '<Terminal>';

  checkRuleRefs(_grammarMap: Map<string, Clause>): void {
    // Terminals have no references to check.
  }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Matches a literal string.
 */
export class Str extends Terminal {
  readonly text: string;

  constructor(text: string) {
    super();
    this.text = text;
  }

  match(parser: Parser, pos: number, _bound?: Clause): MatchResult {
    if (pos + this.text.length > parser.input.length) return mismatch;
    for (let i = 0; i < this.text.length; i++) {
      if (parser.input.charCodeAt(pos + i) !== this.text.charCodeAt(i)) {
        return mismatch;
      }
    }
    return new Match(this, pos, this.text.length);
  }

  toString(): string {
    return `"${escapeString(this.text)}"`;
  }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Matches a single character.
 */
export class Char extends Terminal {
  readonly char: string;

  constructor(char: string) {
    super();
    if (char.length !== 1) {
      throw new Error('Char must be a single character');
    }
    this.char = char;
  }

  match(parser: Parser, pos: number, _bound?: Clause): MatchResult {
    if (pos + this.char.length > parser.input.length) return mismatch;
    for (let i = 0; i < this.char.length; i++) {
      if (parser.input.charCodeAt(pos + i) !== this.char.charCodeAt(i)) {
        return mismatch;
      }
    }
    return new Match(this, pos, this.char.length);
  }

  toString(): string {
    return `'${escapeString(this.char)}'`;
  }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Matches a single character in a set of character ranges.
 *
 * Supports multiple ranges and an optional inversion flag for negated character
 * classes like `[^a-zA-Z0-9]`.
 */
export class CharSet extends Terminal {
  /** List of character ranges as [lo, hi] code unit pairs (inclusive). */
  readonly ranges: ReadonlyArray<readonly [number, number]>;

  /** If true, matches any character NOT in the set. */
  readonly inverted: boolean;

  constructor(ranges: ReadonlyArray<readonly [number, number]>, inverted: boolean = false) {
    super();
    this.ranges = ranges;
    this.inverted = inverted;
  }

  /** Convenience factory for a single character range. */
  static range(lo: string, hi: string): CharSet {
    return new CharSet([[lo.charCodeAt(0), hi.charCodeAt(0)]], false);
  }

  /** Convenience factory for a single character. */
  static char(c: string): CharSet {
    const cp = c.charCodeAt(0);
    return new CharSet([[cp, cp]], false);
  }

  /** Convenience factory for a negated single character range. */
  static notRange(lo: string, hi: string): CharSet {
    return new CharSet([[lo.charCodeAt(0), hi.charCodeAt(0)]], true);
  }

  match(parser: Parser, pos: number, _bound?: Clause): MatchResult {
    if (pos >= parser.input.length) return mismatch;
    const c = parser.input.charCodeAt(pos);

    let inSet = false;
    for (const [lo, hi] of this.ranges) {
      if (c >= lo && c <= hi) {
        inSet = true;
        break;
      }
    }

    if (this.inverted ? !inSet : inSet) {
      return new Match(this, pos, 1);
    }
    return mismatch;
  }

  toString(): string {
    const parts: string[] = ['['];
    if (this.inverted) parts.push('^');
    for (const [lo, hi] of this.ranges) {
      if (lo === hi) {
        parts.push(escapeString(String.fromCharCode(lo)));
      } else {
        parts.push(escapeString(String.fromCharCode(lo)));
        parts.push('-');
        parts.push(escapeString(String.fromCharCode(hi)));
      }
    }
    parts.push(']');
    return parts.join('');
  }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Matches any single character.
 */
export class AnyChar extends Terminal {
  match(parser: Parser, pos: number, _bound?: Clause): MatchResult {
    if (pos >= parser.input.length) return mismatch;
    return new Match(this, pos, 1);
  }

  toString(): string {
    return '.';
  }
}

// -----------------------------------------------------------------------------------------------------------------

/**
 * Matches nothing - always succeeds without consuming any input.
 */
export class Nothing extends Terminal {
  match(_parser: Parser, pos: number, _bound?: Clause): MatchResult {
    return new Match(this, pos, 0);
  }

  toString(): string {
    return '()';
  }
}

File: python/squirrelparser/terminals.py
"""Terminal clauses for the parser."""

from __future__ import annotations
from abc import ABC
from typing import TYPE_CHECKING

from .clause import Clause
from .match_result import Match, mismatch, MatchResult
from .utils import escape_string

if TYPE_CHECKING:
    from .parser import Parser


class Terminal(Clause, ABC):
    """Abstract base class for terminal clauses."""

    # The AST/CST node label for terminals.
    node_label: str = '<Terminal>'

    def check_rule_refs(self, grammar_map: dict[str, Clause]) -> None:
        # Terminals have no references to check.
        pass


# ------------------------------------------------------------------------------------------------------------------


class Str(Terminal):
    """Matches a literal string."""

    __slots__ = ('text',)

    def __init__(self, text: str) -> None:
        self.text = text

    def match(self, parser: Parser, pos: int, *, bound: Clause | None = None) -> MatchResult:
        if pos + len(self.text) > len(parser.input):
            return mismatch
        for i in range(len(self.text)):
            if parser.input[pos + i] != self.text[i]:
                return mismatch
        return Match(self, pos, len(self.text))

    def __repr__(self) -> str:
        return f'"{escape_string(self.text)}"'


# ------------------------------------------------------------------------------------------------------------------


class Char(Terminal):
    """Matches a single character."""

    __slots__ = ('char',)

    def __init__(self, char: str) -> None:
        if len(char) != 1:
            raise ValueError('Char must be a single character')
        self.char = char

    def match(self, parser: Parser, pos: int, *, bound: Clause | None = None) -> MatchResult:
        if pos + len(self.char) > len(parser.input):
            return mismatch
        for i in range(len(self.char)):
            if parser.input[pos + i] != self.char[i]:
                return mismatch
        return Match(self, pos, len(self.char))

    def __repr__(self) -> str:
        return f"'{escape_string(self.char)}'"


# ------------------------------------------------------------------------------------------------------------------


class CharSet(Terminal):
    """
    Matches a single character in a set of character ranges.

    Supports multiple ranges and an optional inversion flag for negated character
    classes like `[^a-zA-Z0-9]`.
    """

    __slots__ = ('ranges', 'inverted')

    def __init__(self, ranges: list[tuple[int, int]], *, inverted: bool = False) -> None:
        """Create a CharSet from a list of (lo, hi) code unit ranges (inclusive)."""
        self.ranges = ranges
        self.inverted = inverted

    @classmethod
    def range(cls, lo: str, hi: str) -> CharSet:
        """Convenience factory for a single character range."""
        return cls([(ord(lo), ord(hi))], inverted=False)

    @classmethod
    def char(cls, c: str) -> CharSet:
        """Convenience factory for a single character."""
        cp = ord(c)
        return cls([(cp, cp)], inverted=False)

    @classmethod
    def not_range(cls, lo: str, hi: str) -> CharSet:
        """Convenience factory for a negated single character range."""
        return cls([(ord(lo), ord(hi))], inverted=True)

    def match(self, parser: Parser, pos: int, *, bound: Clause | None = None) -> MatchResult:
        if pos >= len(parser.input):
            return mismatch
        c = ord(parser.input[pos])

        in_set = False
        for lo, hi in self.ranges:
            if lo <= c <= hi:
                in_set = True
                break

        if (not in_set) if self.inverted else in_set:
            return Match(self, pos, 1)
        return mismatch

    def __repr__(self) -> str:
        parts: list[str] = ['[']
        if self.inverted:
            parts.append('^')
        for lo, hi in self.ranges:
            if lo == hi:
                parts.append(escape_string(chr(lo)))
            else:
                parts.append(escape_string(chr(lo)))
                parts.append('-')
                parts.append(escape_string(chr(hi)))
        parts.append(']')
        return ''.join(parts)


# ------------------------------------------------------------------------------------------------------------------


class AnyChar(Terminal):
    """Matches any single character."""

    def match(self, parser: Parser, pos: int, *, bound: Clause | None = None) -> MatchResult:
        if pos >= len(parser.input):
            return mismatch
        return Match(self, pos, 1)

    def __repr__(self) -> str:
        return '.'


# ------------------------------------------------------------------------------------------------------------------


class Nothing(Terminal):
    """Matches nothing - always succeeds without consuming any input."""

    def match(self, parser: Parser, pos: int, *, bound: Clause | None = None) -> MatchResult:
        return Match(self, pos, 0)

    def __repr__(self) -> str:
        return '()'
